astropy/table/index.py

```diff
@@ -0,0 +1,781 @@
+    ----------
+    table : `Table`
+        Input table
+    table_copy : `Table`
+        Subset of the columns in the table argument
+    '''
+    cols = set(table_copy.columns)
+    indices = set()
+    for column in cols:
+        for index in table[column].indices:
+            if set([x.info.name for x in index.columns]) == cols:
+                return index
+    return None
+
+class _IndexModeContext(object):
+    '''
+    A context manager that allows for special indexing modes, which
+    are intended to improve performance. Currently the allowed modes
+    are "freeze", in which indices are not modified upon column modification,
+    "copy_on_getitem", in which indices are copied upon column slicing,
+    and "discard_on_copy", in which indices are discarded upon table
+    copying/slicing.
+    '''
+
+    def __init__(self, table, mode):
+        '''
+        Parameters
+        ----------
+        table : Table
+            The table to which the mode should be applied
+        mode : str
+            Either 'freeze', 'copy_on_getitem', or 'discard_on_copy'.
+            In 'discard_on_copy' mode,
+            indices are not copied whenever columns or tables are copied.
+            In 'freeze' mode, indices are not modified whenever columns are
+            modified; at the exit of the context, indices refresh themselves
+            based on column values. This mode is intended for scenarios in
+            which one intends to make many additions or modifications in an
+            indexed column.
+            In 'copy_on_getitem' mode, indices are copied when taking column
+            slices as well as table slices, so col[i0:i1] will preserve
+            indices.
+        '''
+        self.table = table
+        self.mode = mode
+        if mode not in ('freeze', 'discard_on_copy', 'copy_on_getitem'):
+            raise ValueError("Expected a mode of either 'freeze', "
+                             "'discard_on_copy', or 'copy_on_getitem', got "
+                             "'{0}'".format(mode))
+
+    def __enter__(self):
+        from .column import (Column, MaskedColumn, _GetitemColumn,
+                             _GetitemMaskedColumn)
+
+        if self.mode == 'discard_on_copy':
+            self.table._copy_indices = False
+        elif self.mode == 'copy_on_getitem':
+            # we need to change column classes temporarily rather
+            # than adjusting instance methods, since new-style
+            # classes refer directly to the class for special
+            # methods
+            self.columns = []
+            self.masked_columns = []
+
+            for col in self.table.columns.values():
+                if isinstance(col, Column):
+                    col.__class__ = _GetitemColumn
+                    self.columns.append(col)
+                elif isinstance(col, MaskedColumn):
+                    col.__class__ = _GetitemMaskedColumn
+                    self.masked_columns.append(col)
```

Comment:
Not that this isn't fine for now, but I also think #4075 removes the need for this temporary class switcheroo, and instead can just tweak the appropriate flags to control the behavior.